### DISCUSSION TITLE
dwc2 support

### DISCUSSION THREAD

**Discussion by cleverca22** [hathach/tinyusb#2856] [General]
### Related area

hcd

### Hardware specification

pi0-pi3

@hathach i notice that your working on a dwc2 hcd, and i thought id share the code ive written to do the same thing
https://github.com/librerpi/lk-overlay/blob/master/platform/bcm28xx/dwc2/dwc2.c

ive got bulk endpoints working, and its able to read from mass-storage, and do dhcp/tftp with LWIP and the built-in nic on an rpi

### I have checked existing issues, dicussion and documentation

- [X] I confirm I have checked existing issues, dicussion and documentation.

**Comment by hathach** [comment]
@cleverca22 ah thank you very much, I am indeed working on hcd driver for dwc2, and able to get pass some control transfer. I will be sure to take a peek at your implementation along the port.

**Comment by cleverca22** [comment]
ive got the rpi ethernet working on both the VPU and aarch64 sides now, but there are some race conditions present that only show up under the aarch64 build, adding a 100ms sleep before each control transfer for the NIC gets things working

edit:
and after some time to think on it, i'm starting to suspect the cache is non-coherent
even if i invalidate a cache line, because [random dma buffers](https://github.com/librerpi/lk-overlay/blob/70d0989459bc3bd5994416b84c20501b0d0ee981/lib/rpi-usb-nic/nic.c#L87) are stack allocated, the cpu can pull them back in before the dma occurs, and then things get very wonky

edit2:
yeah, definitely feels like cache problems
control-out didnt need the sleep, because the existing cache-clean is enough for dma to see the result
control-in worked better when i moved the rx buffer to a dedicated cache line region, my HCD already has a cache-evict
the problem, is that the variable was in the stack, and random activity in the thread un-evicted it, before the DMA could occur
